pkg/registry/apis/dashboard/authorizer.go

```diff
@@ -0,0 +1,104 @@
+
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+
+	"github.com/grafana/authlib/types"
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/dashboards"
+	"github.com/grafana/grafana/pkg/services/guardian"
+	"github.com/grafana/grafana/pkg/storage/legacysql/dualwrite"
+)
+
+func GetAuthorizer(dashboardService dashboards.DashboardService, dualWriter dualwrite.Service, l log.Logger) authorizer.Authorizer {
+	return authorizer.AuthorizerFunc(
+		func(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+			// Note that we will return Allow more than expected.
+			// This is because we do NOT want to hit the RoleAuthorizer that would be evaluated afterwards.
+
+			// Check if we're reading from legacy dashboards and folders
+			// This authorizer is only used for mode 0 to 2. Mode 3 onwards, unified storage handles access control.
+			isReadingLegacy := dualwrite.IsReadingLegacyDashboardsAndFolders(ctx, dualWriter)
+			if !isReadingLegacy {
+				return authorizer.DecisionAllow, "relying on unified storage for access control", nil
+			}
+
+			// Use the standard authorizer
+			if !attr.IsResourceRequest() {
+				// TODO: When is this used?
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+
+			user, err := identity.GetRequester(ctx)
+			if err != nil {
+				return authorizer.DecisionDeny, "", err
+			}
+
+			ns := attr.GetNamespace()
+			if ns == "" {
+				return authorizer.DecisionDeny, "expected namespace", nil
+			}
+
+			info, err := types.ParseNamespace(attr.GetNamespace())
+			if err != nil {
+				return authorizer.DecisionDeny, "error reading org from namespace", err
+			}
+
+			// Validate organization access before we possibly step out here.
+			if user.GetOrgID() != info.OrgID {
+				return authorizer.DecisionDeny, "org mismatch", dashboards.ErrUserIsNotSignedInToOrg
+			}
+
+			if attr.GetVerb() == "create" || attr.GetVerb() == "list" {
+				// Permissions will be handled downstream
+				return authorizer.DecisionAllow, "", nil
+			}
+
+			// Allow search and list requests
+			if attr.GetResource() == "search" || attr.GetName() == "" {
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+
+			// expensive path to lookup permissions for a single dashboard
+			dto, err := dashboardService.GetDashboard(ctx, &dashboards.GetDashboardQuery{
+				UID:   attr.GetName(),
+				OrgID: info.OrgID,
+			})
+			if err != nil {
+				return authorizer.DecisionDeny, "error loading dashboard", err
+			}
+
+			ok := false
+			guardian, err := guardian.NewByDashboard(ctx, dto, info.OrgID, user)
+			if err != nil {
+				return authorizer.DecisionDeny, "", err
+			}
+
+			switch attr.GetVerb() {
+			case "get":
+				ok, err = guardian.CanView()
+				if !ok || err != nil {
+					return authorizer.DecisionDeny, "can not view dashboard", err
+				}
+			case "update":
+				fallthrough
+			case "patch":
```

Comment:
```suggestion
			case "update", "patch":
```
(I did not know this was possible until recently)

Reply:
Me either, thanks!